contributors/design-proposals/storage/volume-topology-scheduling.md

```diff
@@ -266,41 +304,329 @@ const (
 
 This approach allows us to introduce the new binding behavior gradually and to
 be able to maintain backwards compatibility without deprecation of previous
-behavior.  However, it has a few downsides:
+behavior. However, it has a few downsides:
 * StorageClass will be required to get the new binding behavior, even if dynamic
 provisioning is not used (in the case of local storage).
-* We have to maintain two different paths for volume binding.
+* We have to maintain two different code paths for volume binding.
 * We will be depending on the storage admin to correctly configure the
 StorageClasses for the volume types that need the new binding behavior.
 * User experience can be confusing because PVCs could have different binding
 behavior depending on the StorageClass configuration.  We will mitigate this by
 adding a new PVC event to indicate if binding will follow the new behavior.
 
-### Integrating binding with scheduling
-For the alpha phase, the focus is on static provisioning of PVs to support
-persistent local storage.
 
+## Dynamic Provisioning with Topology
+To make dynamic provisioning aware of pod scheduling decisions, delayed volume
+binding must also be enabled. The scheduler will pass its selected node to the
+dynamic provisioner, and the provisioner will create a volume in the topology
+domain that the selected node is part of. The domain depends on the volume
+plugin. Zonal volume plugins will create the volume in the zone where the
+selected node is in. The local volume plugin will create the volume on the
+selected node.
+
+### End to End Zonal Example
+This is an example of the most common use case for provisioning zonal volumes.
+For this use case, the user's specs are unchanged. Only one change
+to the StorageClass is needed to enable delayed volume binding.
+
+1. Admin sets up StorageClass, setting up delayed volume binding.
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: standard
+provisioner: kubernetes.io/gce-pd
+volumeBindingMode: WaitForFirstConsumer
+parameters:
+  type: pd-standard
+```
+2. Admin launches provisioner.  For in-tree plugins, nothing needs to be done.
+3. User creates PVC. Nothing changes in the spec, although now the PVC won't be
+   immediately bound.
+```
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: my-pvc
+spec:
+  storageClassName: standard
+  accessModes:
+  - ReadWriteOnce
+  resources:
+    requests:
+      storage: 100Gi
+```
+4. User creates Pod. Nothing changes in the spec.
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: my-pod
+spec:
+  containers:
+  ...
+  volumes:
+  - name: my-vol
+    persistentVolumeClaim:
+      claimName: my-pvc
+```
+5. Scheduler picks a node that can satisfy the Pod and passes it to the
```

Comment:
@msau42 - Can the scheduler pass the "suggested" nodes where it can be scheduled? What if the node that is picked by the scheduler can't be used by the dynamic provisioner to create a PV. 

In the next step (6), how about the dynamic provisioner create a PV and set nodeAffinity (list of nodes, racks, zones) where it can be accessed and the Pod get scheduled on one of those nodes. This could tie into the way, static PVs provisioned and associated with nodeAffinity works.

Reply:
> What if the node that is picked by the scheduler can't be used by the dynamic provisioner to create a PV.

I think the topology and capacity limitation of the provisioner should have been specified in the Storageclass.

> In the next step (6), how about the dynamic provisioner create a PV and set nodeAffinity (list of nodes, racks, zones) where it can be accessed and the Pod get scheduled on one of those nodes.

This way, there is still a chance that the provisioned pv is conflict to the scheduled node.
I think the main purpose of the feature is to reconcile pod scheduling with volume binding/provisioning,

Reply:
The general scheduler design requires that we have all the scheduler predicates and priorities we need so that the scheduler can make the right decision.  If there is some timing issue which causes the scheduler's view to be out of date, provisioning needs to fail, and the scheduler has to retry.  Eventually, the scheduler's view of the world has to be corrected and it will make the right choice.

If provisioning doesn't fail and it ends up provisioning in a topology domain that the scheduler didn't recommend, in the best case, the scheduler can still find a node that can satisfy the PV.  In the worst case, the pod will fail scheduling and stay pending.

Step 6 will work exactly like as you mentioned.  The PV provisioned must also specify the nodeAffinity, and on pod restarts, it will follow the same bound PVC flow.

Reply:
Thanks @msau42 @laishiquan. Having the topology information to the StorageClass and using that as a scheduler predicate might just work.

Reply:
Can you clarify the use case you are trying to support?  This proposal adds 2 ways to influence scheduling for dynamic provisioning:

1. Admin can restrict topology via AllowedTopologies
2. Provisioner can restrict topology via StorageClassCapacity

Reply:
Do the nodes where the StoragePool exist map to the volume io storage pods?  Also, can a node serve multiple storage pools?

I'm trying to see if you could use the storage class capacity reporting instead of having the admin set allowedTopologies, because that will open up the possibility that the admin will configure it incorrectly.  But the capacity reporting scheme may not work if you can have a node serve multiple storage pools because you cannot have multiple label values for a single key.

Reply:
One more question. Is it a requirement that the application pod is on the same node where the storagepool is in?

Reply:
Hm IIUC, you will have one storageclass per storage pool.  In that case, then capacity reporting won't help, unless the provisioner does something like report some fake capacity for each node in the storagepool (and by default 0 capacity for the rest)

Reply:
A node can server multiple StoragePools depending on the available storage and other resources on the node. 

Since the communication is like this: application pod -> volume io controller pod (stateless) -> storagepool (storage io pod), it is not a hard requirement to have the application pod on the same node as the storagepool, but it is a preferred way. 

@msau42 it is a good point though about the possibility that the admin may configure incorrectly, to circumvent this - we could use an operator to handle the updates to the allowedToplogies. 

Is the the allowedTopologies an editable parameter on the StorageClass object?

Reply:
We could make it mutable. I don't see a reason not to.  Do you see a use case where the admin would want to configure some policy that is different from the operator though?  Maybe they would still want to allow more nodes outside of the storagepool because they are node resource constrained.  

It sounds like for your use case, topology preferences would be better than requirements.

Reply:
Originally, the allowedTopologies was intended to be set by the user, hence, having it in the "spec".  And then capacity is intended to be reported by the system, therefore it's in the "status".  The use case you bring up doesn't fit quite nicely into this model. I need to think about it more

Reply:
@kmova I pondered some more about your use case and I'm not actually sure you need to influence topology at all during provisioning.  Let me know if this makes any sense:

You have one storageclass per storagepool.  A user requests a specific storageclass ie a specific storagepool.  Since it's not a requirement that pods land on the same node where a storagepool spans, you should always have provisioning succeed even if it results in suboptimal placement.  After provisioning, you can specify your topology preferences in the PV.nodeaffinity (note, preferences are not supported today and would need to be added), and the scheduler can use the preferences to make the final node choice.